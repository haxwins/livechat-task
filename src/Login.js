import React, { Component } from 'react';
import { connect } from 'react-redux';
import styled, { keyframes } from 'styled-components';
import { login, logout } from './loginActions';

const emailTest = /^(([^<>()\]\\.,;:\s@"]+(\.[^<>()\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
const passwordTest = /^(?=.*[a-z])(?=.*[A-Z])(?=.*\d)[a-zA-Z\d\D]{6,}$/;

class Login extends Component {
  state = {
    emailValue: '',
    passwordValue: '',
    isPasswordValid: true,
    isEmailValid: true,
  }
 
  onChangeEmail = (e) =>{
    this.setState({emailValue: e.target.value});
  }
  onChangePassword = (e) =>{
    this.setState({passwordValue: e.target.value});
  }

  handleSubmit = () => {
    this.setState({
      isEmailValid: !!emailTest.exec(this.state.emailValue.toLowerCase()),
      isPasswordValid: !!passwordTest.exec(this.state.passwordValue), 
    }, () => {
      if (this.state.isEmailValid && this.state.isPasswordValid) {
        this.props.loginAction(
          this.state.emailValue,
          this.state.passwordValue
        )
      }
    })
  } 

  render() {
    if (this.props.isLoginValid) {
      return (
        <SuccesfullMessage>
          <p>login succesfull</p>
          <Button onClick={this.props.logoutAction}>Logout</Button>
        </SuccesfullMessage>
      )
    }

    return (
      <div>
        {this.props.isLoading && <Loader />}
      <Container>
        <LoginContainer>
          <Input notValid={!this.state.isEmailValid} placeholder="email" value={this.state.emailValue} onChange={this.onChangeEmail} />
          {!this.state.isEmailValid && <ErrorMessage message="invalid email" />}
          <Input notValid={!this.state.isPasswordValid} placeholder="password" value={this.state.passwordValue} type="password" onChange={this.onChangePassword} />
          {!this.state.isPasswordValid && <ErrorMessage message="invalid password" />}
          <Remember label="remember me" />
          <Button onClick={this.handleSubmit}>Login</Button>
          {this.props.errorMessage && <AuthorizationError message={this.props.errorMessage} />}
        </LoginContainer>
      </Container>
      </div>
    );
  }
}

const mapStateToProps = (state) => {
  return {
    ...state,
  };
}

const mapDispatchToProps = (dispatch) => {
  return {
    loginAction: (email, password)=>dispatch(login(email, password)),
    logoutAction: ()=>dispatch(logout())
  }
}

export default connect(mapStateToProps,mapDispatchToProps)(Login)

export const Loader = () => (
  <LoaderContainer>
    <LoaderAnimation></LoaderAnimation>
  </LoaderContainer>
)

export const Remember = ({ label }) => (
  <RememberBoxContainer>
    <RememberLabel>{label}</RememberLabel>
    <RememberCheckBox type="checkbox" />
  </RememberBoxContainer>
)

export const ErrorMessage = ({ message }) => (
  <div>
    <InvalidCredential>{message}</InvalidCredential>
  </div>
)

export const AuthorizationError = ({ message }) => (
  <div>
    <AuthorizationMessage>{message}</AuthorizationMessage>
  </div>
)

export const rotate360 = keyframes`
  from{
    transform: rotate(0deg);
  }
  to{
    transform: rotate(360deg);
  }
`
export const LoaderAnimation = styled.div`
  border: 16px solid #f3f3f3;
  border-top: 16px solid #3498db;
  border-radius: 50%;
  width: 80px;
  height: 80px;
  animation: ${rotate360} 2s linear infinite;
`
export const LoaderContainer = styled.div`
  color: white;
  height: 100vh;
  width: 100vw;
  background-color: rgba(0, 0, 0, 0.5);
  display: flex;
  justify-content: center;
  align-items: center;
  position: absolute;
  z-index: 5;
`
export const Input = styled.input`
  width: 14rem;
  margin: 0.5rem 0 1rem 0;
  padding: 0.5rem;
  color: #69b7eb;
  background: white;
  border: 2px solid; 
  border-color: ${props => props.notValid ? ' 	rgb(255, 0, 51, 0.8)' : 'rgba(0, 0, 0, 0)'};
  border-radius: 6px;
`
export const InvalidCredential = styled.p`
  margin:-1rem 0 0 0.5rem;
  font-size: 0.8rem;
  color: rgb(255, 0, 51, 0.8); 
  position: absolute;
`
export const Container = styled.div`
  font-size: 1.2rem;
  padding-top: 160px;
  display: flex;
  flex-direction: column;
  align-items: center;
  color:rgb(217, 240, 255);
`
export const LoginContainer = styled.div`
  display: flex;
  flex-direction: column;
  margin: 0 0.5rem 0 0.5rem;
`
export const Button = styled.div`
  text-align: center;
  width: 60%;
  color: #69b7eb;
  transform: scale(1, 1);
  transition: transform 0.5s ease;
  cursor: pointer;
  background-color: white;
  border: none;
  border-radius: 6px;
  margin: 0.5rem 0 0 0;
  padding: 0.3rem 1rem 0.3rem 1rem;
  align-self: center;
  :hover{
    transform: scale(1.2, 1.2);
  }
`
export const RememberLabel = styled.p`
  font-size: 16px;
  margin-left: 1.5rem;
  display: inline;
`
export const RememberCheckBox = styled.input`
  margin-left: 1rem;
`
export const RememberBoxContainer = styled.div`
  display: flex;
  justify-content: flex-end;
  flex-direction: row;
  padding-right: 0.5rem;
  align-items: center;
`
export const SuccesfullMessage = styled.div`
  font-size: 1.2rem;
  display: flex;
  align-items: center;
  flex-direction: column;
  color: white;
  width: 400px;
  margin:auto;
  padding-top: 10rem;
`
export const AuthorizationMessage = styled.div`
  margin-top: 1rem;
  font-size: 1rem; 
  color: white;
  text-align: center;
`