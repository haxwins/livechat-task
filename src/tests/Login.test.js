import React, { Component } from 'react';
import ReactDOM from 'react-dom';
import styled from 'styled-components'
import 'jest-styled-components'
import ShallowRenderer from 'react-test-renderer/shallow';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import Login from '../Login.js';
import {
  Loader,
  Remember,
  ErrorMessage,
  AuthorizationError,
  rotate360,
  LoaderAnimation,
  LoaderContainer,
  Input,
  InvalidCredential,
  Container,
  LoginContainer, 
  Button, 
  RememberLabel,
  RememberCheckBox,
  RememberBoxContainer,
  SuccesfullMessage,
  AuthorizationMessage
} from '../Login.js';

const middleweres = [];
const mockStore = configureStore(middleweres);

test('testing whole component with custom state', ()=>{
  const initialState = {
    isLoginValid: false,
    loginMessage: '',
    isLoading: false,
    errorMessage: '',
    }
  const store = mockStore(initialState);

  const renderer = new ShallowRenderer();
  renderer.render(<Login store={store}/>);
  const result = renderer.getRenderOutput();
  expect(result).toMatchSnapshot();
}) 

const components = [
    {   name: 'Button', component: Button},
    {   name: 'Remember', component: Remember},
    {   name: 'ErrorMessage', component: ErrorMessage},
    {   name: 'AuthorizationError', component: AuthorizationError},
    {   name: 'rotate360', component: rotate360},
    {   name: 'LoaderAnimation', component: LoaderAnimation},
    {   name: 'LoaderContainer', component: LoaderContainer},
    {   name: 'Input', component: Input},
    {   name: 'InvalidCredential', component: InvalidCredential},
    {   name: 'Container', component: Container},
    {   name: 'LoginContainer', component: LoginContainer},
    {   name: 'Button', component: Button},
    {   name: 'RememberLabel', component: RememberLabel},
    {   name: 'RememberCheckBox', component: RememberCheckBox},
    {   name: 'RememberBoxContainer', component: RememberBoxContainer},
    {   name: 'SuccesfullMessage', component: SuccesfullMessage},
    {   name: 'AuthorizationMessage', component: AuthorizationMessage},
]

components.forEach((element) => {
    test(element.name, () => {
        const Temp = element.component;
        const tree = renderer.create(<Temp />).toJSON()
        expect(tree).toMatchSnapshot()
    })
})

test('rendering component when login is succesfull', ()=>{
    const initialState = {
      isLoginValid: true,
      loginMessage: '',
      isLoading: false,
      errorMessage: '',
      }
    const store = mockStore(initialState);
  
    const tree = renderer.create(<Login store={store}/>).toJSON()
    expect(tree).toMatchSnapshot();
  }) 