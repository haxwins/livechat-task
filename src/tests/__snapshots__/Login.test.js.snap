// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthorizationError 1`] = `
.c0 {
  margin-top: 1rem;
  font-size: 1rem;
  color: white;
  text-align: center;
}

<div>
  <div
    className="c0"
  />
</div>
`;

exports[`AuthorizationMessage 1`] = `
.c0 {
  margin-top: 1rem;
  font-size: 1rem;
  color: white;
  text-align: center;
}

<div
  className="c0"
/>
`;

exports[`Button 1`] = `
.c0 {
  text-align: center;
  width: 60%;
  color: #69b7eb;
  -webkit-transform: scale(1,1);
  -ms-transform: scale(1,1);
  transform: scale(1,1);
  -webkit-transition: -webkit-transform 0.5s ease;
  -webkit-transition: transform 0.5s ease;
  transition: transform 0.5s ease;
  cursor: pointer;
  background-color: white;
  border: none;
  border-radius: 6px;
  margin: 0.5rem 0 0 0;
  padding: 0.3rem 1rem 0.3rem 1rem;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
}

.c0:hover {
  -webkit-transform: scale(1.2,1.2);
  -ms-transform: scale(1.2,1.2);
  transform: scale(1.2,1.2);
}

<div
  className="c0"
/>
`;

exports[`Button 2`] = `
.c0 {
  text-align: center;
  width: 60%;
  color: #69b7eb;
  -webkit-transform: scale(1,1);
  -ms-transform: scale(1,1);
  transform: scale(1,1);
  -webkit-transition: -webkit-transform 0.5s ease;
  -webkit-transition: transform 0.5s ease;
  transition: transform 0.5s ease;
  cursor: pointer;
  background-color: white;
  border: none;
  border-radius: 6px;
  margin: 0.5rem 0 0 0;
  padding: 0.3rem 1rem 0.3rem 1rem;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
}

.c0:hover {
  -webkit-transform: scale(1.2,1.2);
  -ms-transform: scale(1.2,1.2);
  transform: scale(1.2,1.2);
}

<div
  className="c0"
/>
`;

exports[`Container 1`] = `
.c0 {
  font-size: 1.2rem;
  padding-top: 160px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  color: rgb(217,240,255);
}

<div
  className="c0"
/>
`;

exports[`ErrorMessage 1`] = `
.c0 {
  margin: -1rem 0 0 0.5rem;
  font-size: 0.8rem;
  color: rgb(255,0,51,0.8);
  position: absolute;
}

<div>
  <p
    className="c0"
  />
</div>
`;

exports[`Input 1`] = `
.c0 {
  width: 14rem;
  margin: 0.5rem 0 1rem 0;
  padding: 0.5rem;
  color: #69b7eb;
  background: white;
  border: 2px solid;
  border-color: rgba(0,0,0,0);
  border-radius: 6px;
}

<input
  className="c0"
/>
`;

exports[`InvalidCredential 1`] = `
.c0 {
  margin: -1rem 0 0 0.5rem;
  font-size: 0.8rem;
  color: rgb(255,0,51,0.8);
  position: absolute;
}

<p
  className="c0"
/>
`;

exports[`LoaderAnimation 1`] = `
.c0 {
  border: 16px solid #f3f3f3;
  border-top: 16px solid #3498db;
  border-radius: 50%;
  width: 80px;
  height: 80px;
  -webkit-animation: iVXCSc 2s linear infinite;
  animation: iVXCSc 2s linear infinite;
}

<div
  className="c0"
/>
`;

exports[`LoaderContainer 1`] = `
.c0 {
  color: white;
  height: 100vh;
  width: 100vw;
  background-color: rgba(0,0,0,0.5);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  position: absolute;
  z-index: 5;
}

<div
  className="c0"
/>
`;

exports[`LoginContainer 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  margin: 0 0.5rem 0 0.5rem;
}

<div
  className="c0"
/>
`;

exports[`Remember 1`] = `
.c1 {
  font-size: 16px;
  margin-left: 1.5rem;
  display: inline;
}

.c2 {
  margin-left: 1rem;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  padding-right: 0.5rem;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

<div
  className="c0"
>
  <p
    className="c1"
  />
  <input
    className="c2"
    type="checkbox"
  />
</div>
`;

exports[`RememberBoxContainer 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  padding-right: 0.5rem;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

<div
  className="c0"
/>
`;

exports[`RememberCheckBox 1`] = `
.c0 {
  margin-left: 1rem;
}

<input
  className="c0"
/>
`;

exports[`RememberLabel 1`] = `
.c0 {
  font-size: 16px;
  margin-left: 1.5rem;
  display: inline;
}

<p
  className="c0"
/>
`;

exports[`SuccesfullMessage 1`] = `
.c0 {
  font-size: 1.2rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  color: white;
  width: 400px;
  margin: auto;
  padding-top: 10rem;
}

<div
  className="c0"
/>
`;

exports[`rendering component when login is succesfull 1`] = `
.c1 {
  text-align: center;
  width: 60%;
  color: #69b7eb;
  -webkit-transform: scale(1,1);
  -ms-transform: scale(1,1);
  transform: scale(1,1);
  -webkit-transition: -webkit-transform 0.5s ease;
  -webkit-transition: transform 0.5s ease;
  transition: transform 0.5s ease;
  cursor: pointer;
  background-color: white;
  border: none;
  border-radius: 6px;
  margin: 0.5rem 0 0 0;
  padding: 0.3rem 1rem 0.3rem 1rem;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
}

.c1:hover {
  -webkit-transform: scale(1.2,1.2);
  -ms-transform: scale(1.2,1.2);
  transform: scale(1.2,1.2);
}

.c0 {
  font-size: 1.2rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  color: white;
  width: 400px;
  margin: auto;
  padding-top: 10rem;
}

<div
  className="c0"
>
  <p>
    login succesfull
  </p>
  <div
    className="c1"
    onClick={[Function]}
  >
    Logout
  </div>
</div>
`;

exports[`rotate360 1`] = `<iVXCSc />`;

exports[`testing whole component with custom state 1`] = `
<Login
  errorMessage=""
  isLoading={false}
  isLoginValid={false}
  loginAction={[Function]}
  loginMessage=""
  logoutAction={[Function]}
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
  storeSubscription={
    Subscription {
      "listeners": Object {
        "notify": [Function],
      },
      "onStateChange": [Function],
      "parentSub": undefined,
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "unsubscribe": null,
    }
  }
/>
`;
